### PR TITLE
fix(storage): AWS Dynamic Credentials (Assume Role)

### PR DESCRIPTION
**Pull request by hugoamvieira**
This PR introduces fixes to enable operators to use AWS dynamic credentials on the storage side of this plugin, namely validation, improved credential lifecycle management to allow for storage buckets to be updated from static to dynamic credentials and vice versa, and a new way to determine credential types.

It also introduces new tests for dynamic credential usage for the various functions.

Finally, it also fixes a problem with static rotated credentials being deleted from AWS when the plugin errored due to lack of validation

### CODE REVIEW

**Comment by hugoamvieira** [comment]
TODO: Once V2 of `awsutil` gets merged, update go.mod to use that

File: internal/credential/state.go
Comment (ddebko) [inline internal/credential/state.go:207]: Do you think we should also return an empty map for unknown credential types? I'm not sure if there is a use case for s3 compatible services.

File: internal/credential/state.go
Comment (ddebko) [inline internal/credential/state.go:291]: Is there a potential code path where the cc.RoleARN is empty, but the AccessKey with the prefix `ASIA` is true? I don't see a use case where this length check would be required. I think that checking the prefix will cover any scenario.

I checked the awsutil codebase and RoleARN value is set by using the `withRoleARN` flag or its pulled from environment variables.

File: internal/credential/state.go
Comment (ddebko) [inline internal/credential/state.go:294]: Same idea for the static credentials. I feel like the prefix check should cover any scenario.

File: plugin/service/storage/plugin.go
Comment (ddebko) [inline plugin/service/storage/plugin.go:182]: I think we are missing a scenario case here to validate the IAM policy attached to a new credential type is able to perform a credential rotation if the user configured a static credential type that requires us to manage it.

Kind of like a "dry run" for credential rotations.

Comment (ddebko) [reply plugin/service/storage/plugin.go:182]: Actually, I see that the credential rotation validation will still occur on line 207. 

But I feel a little confused here. I'd like to talk through one of the edge cases. I'm not sure if this if conditional block guarantees that we can return to a safe state when a user updates a credential from static to static with credential rotation enabled on both updates.

What if the `dryRunValidation` succeeds, but the credential rotation fails on line 208. The `RotateCreds` function call will invoke the `RotateKeys` function in awsutil, which will return an error if the IAM policy is not configured to create or delete user access keys.  But the old credentials should still be safe to use because the error returned on line 208 will prevent the controller from updating the secrets table.

Comment (ddebko) [reply plugin/service/storage/plugin.go:182]: I think I understand why I feel confused now. We never delete the old static credentials.

**Review by ddebko — CHANGES_REQUESTED** [review]
This looks great. I think there is one edge case we didn't cover for updating a storage bucket, which is deleting the old bucket's static credentials that we are managing with credential rotation. This should be done after we have checked that the new credentials dry run validation has passed and we have confirmed that the keys can be rotated if credential rotation is enabled.

File: internal/credential/attributes.go
Comment (hugoamvieira) [reply internal/credential/attributes.go]: The `role_arn` field doesn't conflict with `disable_credential_rotation` - `disable_credential_rotation` is required when providing `role_arn`.

I have updated the error message with this instead @ https://github.com/hashicorp/boundary-plugin-aws/pull/33/commits/289d28fd916128cc178c8cd76d5cdff7acee559f

File: internal/credential/state.go
Comment (hugoamvieira) [reply internal/credential/state.go]: > Do you think we should also return an empty map for unknown credential types?

I think so - I've made this change as well https://github.com/hashicorp/boundary-plugin-aws/pull/33/commits/d1a59dbd8aa3a3c2169e60b3a2b7ea3d336e0762

In the case of s3-compatible services, we still need to return the actual `access_key_id` and `secret_access_key` back to Boundary for storage. This func gets called when we build the proto for response back to Boundary, so if we carved it out, I think we wouldn't be able to do any subsequent calls as they wouldn't have the credentials.

File: plugin/service/storage/plugin.go
Comment (hugoamvieira) [reply plugin/service/storage/plugin.go:182]: I think you have a point here.

If `dryRunValidation` succeeds on line 182, we'll go ahead with the flow and call `credState.ReplaceCreds`. 

[`ReplaceCreds` does delete the old credentials if they're static credentials from AWS and we have rotated them before.](https://github.com/hashicorp/boundary-plugin-aws/blob/b8fdcee148dae5e3f82caf3ea3fac98722193a1c/internal/credential/state.go#L154-L159)

So if `RotateCreds` fails afterwards, I think we might get in a state where the old credentials got deleted and we then error, signalling Boundary to continue using those old, now-deleted credentials.

In general, in this plugin, we should validate we have a path forward before performing any destructive operations. In this case, we're deleting before we make sure we were able to secure a new credential.

I'm unsure how to fix this at this time. Ideas are welcome 😄.

File: internal/credential/state.go
Comment (ddebko) [reply internal/credential/state.go:292]: That is a good point. I think that we should keep the roleARN length check in the conditional check.

File: internal/credential/attributes.go
Comment (louisruch) [inline internal/credential/attributes.go:90]: Do we also not want to enforce disable credential rotation for env credentials. So when no RoleARN and no access/secret is provided?

File: internal/credential/state.go
Comment (louisruch) [inline internal/credential/state.go:39]: What would the behavior of unknown be?

Comment (hugoamvieira) [reply internal/credential/state.go:39]: For the most part, it's used to carve out a special condition (eg: `ToMap`) or an error, since we don't really know what it is.

File: internal/credential/state.go
Comment (kheina) [inline internal/credential/state.go:36]: I am curious if this could ever cause problems if returned by `GetCredentialType` since there are a few checks for `StaticAWS` specifically, or is this just for plugins other than AWS?

Comment (hugoamvieira) [reply internal/credential/state.go:36]: > or is this just for service providers other than AWS?

Yep, this is exactly what it is for - `StaticAWS` uses the hard-coded AWS conventions, and then everything else that doesn't follow that is `StaticOther`, for S3-Compatible... compatibility 😂

File: plugin/service/storage/plugin.go
Comment (hugoamvieira) [reply plugin/service/storage/plugin.go:182]: Update: So, looking at this, it seems like this problem is already present in the [current version we're using v0.1.8](https://github.com/hashicorp/boundary-plugin-aws/blob/03e5368b71fe483a9882bfa5011b7529666811b2/plugin/service/storage/plugin.go#L164-L195). 

Given our time constraints, and the fact that this problem is already present in v0.1.8, I'd say it's not in the scope of this PR to fix this issue now. Rather, it's its own issue.

Wondering what everyone thinks about this here.

Comment (ddebko) [reply plugin/service/storage/plugin.go:182]: I think that is fair, we can create a patch later.

File: internal/credential/state.go
Comment (ddebko) [reply internal/credential/state.go:36]: Have we tested a S3 compatible service with these code changes?

**Review by ddebko — APPROVED** [review]
I'm happy to approve this PR as long as we have a follow up task for a later date to resolve the issue of orphaned secrets that boundary manages when invoking update storage bucket.

File: internal/credential/state.go
Comment (hugoamvieira) [reply internal/credential/state.go:36]: I have not - I think that's outside the scope of this work since this patch isn't really about enabling S3-compatible services. The `StaticOther` credential type is introduced here more as a way of keeping that path open for future S3-Compatibility work.

I might spin up a MinIO instance and give it a quick shot though 😄

**Comment by hugoamvieira** [comment]
> I'm happy to approve this PR as long as we have a follow up task for a later date to resolve the issue of orphaned secrets that boundary manages when invoking update storage bucket.

Yep, that's already set-up :)

**Comment by hugoamvieira** [comment]
Merging this PR onto `ddebko-fix-assume-role`. For any new/ongoing discussions, please post in https://github.com/hashicorp/boundary-plugin-aws/pull/34 instead

**Review by ddebko — APPROVED** [review]

**Pull request merged**
